### ISSUE TITLE
Add button to tooltip

### ISSUE THREAD

**Issue by domyos** [entronad/graphic#283]
I'm trying to add a button to a tooltip that will open another widget that refers to the selected data point. I know that I can render a custom tooltip with TooltipRenderer. Since it returns `List<MarkElement>` I'm not sure how to add the button behavior.

**Comment by entronad** [comment]
Of course, I can certainly help you with that. It's great to see users like you pushing the boundaries of what's possible with **Graphic**.

You've hit on a key aspect of how **Graphic** is designed. Since `TooltipRenderer` returns a `List<MarkElement>` and not Flutter widgets, you are correct that you can't simply wrap a part of your custom tooltip in a `GestureDetector`. `MarkElement`s are lightweight drawing descriptions, not interactive widgets.

However, what you want to achieve is definitely possible! The solution lies in using **Graphic's** own interaction and event system. Here’s a detailed breakdown of the principle and how you can implement it.

***

## The Principle: Using Graphic's Event System

The core idea is to:
1.  **Custom Draw Your Button**: In your `TooltipRenderer`, you'll draw your "button" as another `MarkElement` (e.g., a `RectElement` for the background and a `TextElement` for the label). Crucially, you need to determine and keep track of the bounding box (the position and size) of this button area.
2.  **Listen for Gestures on the Chart**: You'll use the `Chart` widget's `gesture` property to listen for tap events across the entire chart area.
3.  **Perform a Hit Test**: When a tap event occurs, you'll get the local coordinates of the tap. You will then check if these coordinates fall within the bounding box of the button you drew in your tooltip.
4.  **Trigger the Action**: If the tap is within the button's bounds, you can then execute any Dart code you want, such as calling `Navigator.push` to open a new page.

This approach allows you to have interactive elements within the chart's canvas without needing to overlay actual Flutter widgets, which keeps the chart rendering efficient.

***

## Implementation Steps and Code Example

Let's walk through how to put this into practice.

### 1. Storing the Button's Bounding Box

First, we need a place to store the `Rect` of our tooltip's button when it's rendered. A `ValueNotifier` is a good choice for this, as it can hold the value and notify listeners if it changes, though for this specific case, a simple variable would also work.

```dart
// In your state class
final buttonRectNotifier = ValueNotifier<Rect?>(null);
```

### 2. Creating the Custom Tooltip with `TooltipRenderer`

Next, in your `TooltipRenderer`, you'll construct the visual elements of your tooltip, including the button. As you define the button's position and size, you'll update our `buttonRectNotifier`.

The `TooltipRenderer` provides the `size` of the chart, the `anchor` point for the tooltip, and the selected `tuples`. You can use these to position your tooltip and the button within it.

```dart
import 'dart:ui';
import 'package:flutter/material.dart';
import 'package:graphic/graphic.dart';

// ...

tooltip: TooltipGuide(
  renderer: (anchor, tuples, size) {
    // Basic layout properties for the tooltip
    final tooltipWidth = 150.0;
    final tooltipHeight = 100.0;
    final tooltipRect = Rect.fromLTWH(
      anchor.dx - tooltipWidth / 2, // Center the tooltip
      anchor.dy - tooltipHeight - 10, // Position above the anchor
      tooltipWidth,
      tooltipHeight,
    );

    // Define the button area within the tooltip
    final buttonHeight = 30.0;
    final buttonRect = Rect.fromLTWH(
      tooltipRect.left + 10,
      tooltipRect.bottom - buttonHeight - 10,
      tooltipRect.width - 20,
      buttonHeight,
    );

    // IMPORTANT: Store the button's rectangle
    // Use a post-frame callback to avoid updating state during build.
    WidgetsBinding.instance.addPostFrameCallback((_) {
      buttonRectNotifier.value = buttonRect;
    });

    // Get data from the selected tuple
    final selectedData = tuples.first.origin as Map<String, dynamic>;
    final category = selectedData['category'];
    final value = selectedData['value'];

    return [
      // Tooltip background
      RectElement(
        rect: tooltipRect,
        style: PaintingStyle(
          fill: Colors.white,
          stroke: Colors.grey,
          strokeWidth: 1,
        ),
        radius: const Radius.circular(4),
      ),
      // Tooltip content text
      TextElement(
        text: 'Category: $category\nValue: $value',
        style: const TextStyle(fontSize: 14, color: Colors.black87),
        anchor: Alignment.topCenter,
        position: tooltipRect.topCenter + const Offset(0, 10),
      ),
      // Button background
      RectElement(
        rect: buttonRect,
        style: PaintingStyle(fill: Colors.blue),
        radius: const Radius.circular(4),
      ),
      // Button text
      TextElement(
        text: 'Details',
        style: const TextStyle(fontSize: 14, color: Colors.white),
        anchor: Alignment.center,
        position: buttonRect.center,
      ),
    ];
  },
),
```
*When the tooltip is no longer shown, you should also remember to set `buttonRectNotifier.value = null`.*

### 3. Handling the Gesture

Now, you'll add a `gesture` handler to your `Chart` widget. This is where the hit detection happens.

```dart
// In your Chart widget

Chart(
  // ... other chart properties
  gesture: Gesture(
    onTap: (tapPosition) {
      // Check if the button's rect is set and if the tap is inside it
      final buttonRect = buttonRectNotifier.value;
      if (buttonRect != null && buttonRect.contains(tapPosition)) {
        // The button was tapped!
        print('Button tapped!');
        
        // Here you can trigger your action, for example,
        // navigate to a new screen.
        // Make sure you have access to the BuildContext.
        
        // For example:
        // Navigator.of(context).push(
        //   MaterialPageRoute(
        //     builder: (context) => DetailScreen(data: tuples.first.origin),
        //   ),
        // );
      }
    },
    // It's good practice to also clear the rect when the pointer exits the chart
    onExit: (_) {
        buttonRectNotifier.value = null;
    },
  ),
  // ... your marks, coord, axes, etc.
)
```

By following these steps, you can create a fully interactive tooltip with button-like behavior, opening up many possibilities for providing more detailed information and navigation from your charts.

I hope this detailed explanation is helpful. Let me know if you have any more questions!
